SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/RPCService.cpp

```diff
@@ -0,0 +1,339 @@
+					{
+						QueueEntry.Value.Sender->OnUpdate(Ctx);
+					}
+				}
+			}
+			break;
+		}
+		case EntityDelta::ADD:
+		{
+			const EntityViewElement& ViewElement = LocalAuthSubView->GetView().FindChecked(Delta.EntityId);
+			for (auto& QueueEntry : Queues)
+			{
+				if (ViewElement.Authority.Contains(QueueEntry.Value.Authority))
+				{
+					QueueEntry.Value.Queue->OnAuthGained(Delta.EntityId, ViewElement);
+					QueueEntry.Value.Sender->OnAuthGained(Delta.EntityId, ViewElement);
+				}
+			}
+		}
+		break;
+		case EntityDelta::REMOVE:
+			for (auto& QueueEntry : Queues)
+			{
+				QueueEntry.Value.Queue->OnAuthLost(Delta.EntityId);
+				QueueEntry.Value.Sender->OnAuthLost(Delta.EntityId);
+			}
+			break;
+		case EntityDelta::TEMPORARILY_REMOVED:
+		{
+			const EntityViewElement& ViewElement = LocalAuthSubView->GetView().FindChecked(Delta.EntityId);
+			for (auto& QueueEntry : Queues)
+			{
+				QueueEntry.Value.Queue->OnAuthLost(Delta.EntityId);
+				QueueEntry.Value.Sender->OnAuthLost(Delta.EntityId);
+				if (ViewElement.Authority.Contains(QueueEntry.Value.Authority))
+				{
+					QueueEntry.Value.Queue->OnAuthGained(Delta.EntityId, ViewElement);
+					QueueEntry.Value.Sender->OnAuthGained(Delta.EntityId, ViewElement);
+				}
+			}
+		}
+		break;
+		default:
+			checkNoEntry();
+			break;
+		}
+	}
+}
+
+void RPCService::AdvanceReceivers()
+{
+	auto HasReceiverAuthority = [](const RPCReceiverDescription& Desc, const EntityViewElement& ViewElement) {
+		return Desc.Authority == 0 || ViewElement.Authority.Contains(Desc.Authority);
+	};
+
+	for (const EntityDelta& Delta : RemoteSubView->GetViewDelta().EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			const EntityViewElement& ViewElement = RemoteSubView->GetView().FindChecked(Delta.EntityId);
+			for (const ComponentChange& Change : Delta.ComponentUpdates)
+			{
+				RPCReadingContext Ctx;
+				Ctx.EntityId = Delta.EntityId;
+				Ctx.ComponentId = Change.ComponentId;
+				Ctx.Update = Change.Update;
+				Ctx.Fields = Schema_GetComponentUpdateFields(Change.Update);
+
+				for (auto& Entry : Receivers)
+				{
+					if (HasReceiverAuthority(Entry.Value, ViewElement)
+						&& Entry.Value.Receiver->GetComponentsToRead().Contains(Ctx.ComponentId))
+					{
+						Entry.Value.Receiver->OnUpdate(Ctx);
+					}
+				}
+			}
+			for (const ComponentChange& Change : Delta.ComponentsRefreshed)
+			{
+				RPCReadingContext Ctx;
+				Ctx.EntityId = Delta.EntityId;
+				Ctx.ComponentId = Change.ComponentId;
+				Ctx.Fields = Schema_GetComponentDataFields(Change.CompleteUpdate.Data);
+
+				for (auto& Entry : Receivers)
+				{
+					if (HasReceiverAuthority(Entry.Value, ViewElement)
+						&& Entry.Value.Receiver->GetComponentsToRead().Contains(Ctx.ComponentId))
+					{
+						Entry.Value.Receiver->OnUpdate(Ctx);
+					}
+				}
+			}
+
+			for (const AuthorityChange& Change : Delta.AuthorityLost)
```

Comment:
I am not sure I understand what is going on here, you are iterating over Changes but not using `Change` at all, in 4 separate loops?

Reply:
Hmm true, I should check the auth value.